### PR TITLE
Undo notification should not show when an action fails #1237

### LINKED ISSUE #1237: Undo notification should not show when an action fails
I noticed that the undo notification still displays when I attempted to run an action which used GitHub's API incorrectly (and thus failed).
The relevant method that should be modified is addAction in UndoController.
screen shot 2016-02-04 at 10 30 47 am

### PR DESCRIPTION
**Pull request by m133225**
Fixes #1237.

### CODE REVIEW

**Comment by m133225** [comment]
Ready for review.

File: src/main/java/undo/UndoController.java [hunk 1]
```diff
@@ -48,28 +48,40 @@ public void undoCallback() {
...
-        action.act(issue).thenApply(success -> showErrorDialogOnFailure(success, issue, action));
-        Notification notification = new Notification(createInfoOcticon(),
-                action.getDescription() + " for #" + issue.getId() + ": " + issue.getTitle(),
-                "Undo", this::undoCallback);
-        notificationController.showNotification(notification);
+        action.act(issue).thenApply(success -> handleActionResult(issue, action, success, false));
     }
 
-    private boolean showErrorDialogOnFailure(Boolean success, TurboIssue issue, Action action) {
+    private boolean handleActionResult(TurboIssue issue, Action<TurboIssue> action, Boolean success,
```
Comment (acruis) [inline src/main/java/undo/UndoController.java]: Do write javadoc comments for this method and `showNotification`, `showErrorDialog`.

**Comment by acruis** [comment]
Any tests for this? I think we can use `MockServer` to test this behaviour.

**Comment by m133225** [comment]
@acruis Ready for review. Added tests and comments.

File: src/main/java/undo/UndoController.java [hunk 2]
```diff
@@ -48,28 +48,61 @@ public void undoCallback() {
...
+     * Possibly shows a notification or an error dialog depending on {@code success} and {@code isUndo}.
+     *
+     * @param issue the TurboIssue acted on
+     * @param action the Action that acted on the issue
+     * @param success whether the action was successful
+     * @param isUndo whether action is an undo
+     * @return {@code success}
+     */
+    private boolean handleActionResult(TurboIssue issue, Action<TurboIssue> action, Boolean success,
+                                           boolean isUndo) {
```
Comment (damithc) [inline src/main/java/undo/UndoController.java]: This indentation is a bit odd.

File: src/test/java/tests/NotificationTests.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import static org.mockito.Mockito.*;
+
+public class NotificationTests {
+    NotificationController notificationController;
+    UndoController undoController;
+    Action<TurboIssue> action;
+
+    @Before
+    public void initComponents() {
+        // required so notification construction won't fail
```
Comment (acruis) [inline src/test/java/tests/NotificationTests.java]: Elaborate more on this in the comments.

**Comment by acruis** [comment]
Good work. Just some more documentation and this should be good to go.

**Comment by m133225** [comment]
Made the appropriate changes!

**Comment by acruis** [comment]
Okay, no more comments from me.

**Pull request merged**